### PR TITLE
Capstone - Redis config

### PR DESCRIPTION
**Pull request by danik-tro**
# redis-config crate
Implemented an Asynchronous Redis Source compatible with `config-rs` crate to extend the sources list provided by `configs-rs`.

### CODE REVIEW

File: .github/issue_template.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+
+* **What is the motivation / use case for changing the behavior?**
+
+
+
+* **Please tell us about your environment:**
+  
+  - Version: 2.0.0-beta.X
+  - Browser: [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]
```
Comment (sokyrko) [inline .github/issue_template.md:33]: Can you run this in browser? 0_o

**Review by sokyrko — CHANGES_REQUESTED** [review]
Also, please add:
- Rust CI (fmt, lint, test)
- crates.io deployment pipeline (automated version bump is nice to have as well).

File: .github/issue_template.md (hunk 1 above)
Comment (danik-tro) [reply .github/issue_template.md]: Copypasta from issue-templates, fixed

File: src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    ///
+    /// Will return `SourceError::EnvVariableNotSet` if the environment variable is not set.
+    pub fn from_env(source_key: SK) -> SourceResult<Self> {
+        let redis_url = std::env::var(REDIS_CONFIG_URL_KEY)
+            .map_err(|_| SourceError::EnvVariableNotSet(REDIS_CONFIG_URL_KEY.into()))?;
+
+        Self::try_new(source_key, redis_url)
+    }
+
+    /// Initialize a new redis source with provided `source_key` and `connnection_info` to Redis.
```
Comment (tiamo) [inline src/lib.rs:93]: typo `connNection_info`

File: src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+            .get_async_connection()
+            .await?
+            .get(self.source_key.clone())
+            .await?;
+
+        if !self.required && data.is_none() {
+            return Ok(Map::new());
+        }
+
+        let Some(data) = data else {
```
Comment (tiamo) [inline src/lib.rs:142]: Can be simplified
```rust
let data = data.ok_or(SourceError::RedisKeyDoesNotExist)?;
```

Comment (danik-tro) [reply src/lib.rs]: Fixed, thanks

File: src/lib.rs (hunk 2 above)
Comment (danik-tro) [reply src/lib.rs]: Fixed

File: examples/redis_key_source.rs [hunk 4]
```diff
@@ -0,0 +1,53 @@
+use config::builder::AsyncState;
+use redis::AsyncCommands;
+use redis_config::RedisSource;
+
+const REDIS_URL: &str = "redis://127.0.0.1:6379";
+const SOURCE_KEY: &str = "application-settings";
```
Comment (pahoma) [inline examples/redis_key_source.rs:6]: Seems it's better to use a flexible configuration file with the ability to use different APP_ENV configs and env variables.

File: src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+const REDIS_CONFIG_URL_KEY: &str = "REDIS_CONFIG_URL";
+
+/// A configuration async source backed up by a Redis.
+///
+/// It supports retrieving configuration in JSON format from Strings and Hashes types in Redis.
+///
+/// # Examples
+///
+/// ```ignore
```
Comment (vni) [inline src/lib.rs:40]: is it possible to un-ignore this doctest?
(I failed to quickly find a solutions for compiling ```let config = get_config().await.unwrap();``` here (the last line), but maybe you will have more time and willingness)

File: examples/redis_key_source.rs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+#[derive(Debug, serde::Serialize, serde::Deserialize, PartialEq, Eq)]
+struct ServerSettings {
+    ttl: i64,
+    path: String,
+    // ...
```
Comment (pahoma) [inline examples/redis_key_source.rs:12]: Typo ?

**Review by vni — COMMENTED** [review]
As for me, looks good.
Please, fix minor issues from other reviewers.

File: src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+mod errors;
+
+pub use errors::{SourceError, SourceResult};
+
+use std::{collections::HashMap, fmt::Debug};
+
+use config::{AsyncSource, ConfigError, Map};
+use redis::AsyncCommands;
+
+const REDIS_CONFIG_URL_KEY: &str = "REDIS_CONFIG_URL";
```
Comment (pahoma) [inline src/lib.rs:32]: Also to config

**Review by pahoma — CHANGES_REQUESTED** [review]

File: examples/redis_key_source.rs (hunk 6 above)
Comment (danik-tro) [reply examples/redis_key_source.rs:21]: Fixed

File: src/lib.rs (hunk 7 above)
Comment (danik-tro) [reply src/lib.rs]: Added an Env source in the examples/ directory

File: examples/redis_key_source.rs (hunk 4 above)
Comment (danik-tro) [reply examples/redis_key_source.rs:7]: Added an Env source in the examples/ directory

File: src/lib.rs (hunk 5 above)
Comment (danik-tro) [reply src/lib.rs]: Fixed, example was incorrect. Changed to no_run for this doc test, to not run integration tests in doc tests.

But you can run it locally with docker

`docker run --name bootcamp_redis -p 6379:6379 -d redis`

**Review by sokyrko — APPROVED** [review]
Nice =)

File: examples/redis_key_source.rs (hunk 4 above)
Comment (pahoma) [reply examples/redis_key_source.rs:7]: Awesome!!!

**Review by pahoma — APPROVED** [review]

**Review by vni — APPROVED** [review]

**Review by TheBestTvarynka — APPROVED** [review]

**Review by tiamo — APPROVED** [review]
👍

**Comment by danik-tro** [comment]
@sokyrko @vni @tiamo @TheBestTvarynka @pahoma 

Thanks a lot for the reviews and comments!

**Pull request merged**